src/db/file.rs

```diff
@@ -138,102 +138,97 @@ pub fn add_path_into_database<P: AsRef<Path>>(conn: &Connection,
-                        key: bucket_path.clone(),
-                        body: Some(content.clone().into()),
-                        content_type: Some(mime.clone()),
-                        ..Default::default()
-                    }).sync();
-                    attempts += 1;
-                    match s3_res {
-                        // we've successfully uploaded the content, so steal it;
-                        // we don't want to put it in the DB
-                        Ok(_) => break None,
-                        // Since s3 was configured, we want to panic on failure to upload.
-                        Err(e) => {
-                            log::error!("failed to upload to {}: {:?}", bucket_path, e);
-                            // Get a new client, in case the old one's connection is stale.
-                            // AWS will kill our connection if it's alive for too long; this avoids
-                            // that preventing us from building the crate entirely.
-                            *client = s3_client().unwrap();
-                            if attempts > 3 {
-                                panic!("failed to upload 3 times, exiting");
-                            } else {
-                                continue;
-                            }
-                        },
-                    }
-                }
             } else {
-                Some(content.clone().into())
-            };
-
-            file_list_with_mimes.push((mime.clone(), file_path.clone()));
-
-            (
-                bucket_path,
-                content,
-                mime,
-            )
+                mime.to_owned()
+            }
         };
 
-        // If AWS credentials are configured, don't insert/update the database
-        if client.is_none() {
+        if let Some(client) = s3_client() {
+            let bucket_path = bucket_path.clone();
+            let content = content.clone();
+            let mime = mime.clone();
+
+            futures.push(client.put_object(PutObjectRequest {
+                bucket: "rust-docs-rs".into(),
+                key: bucket_path.clone(),
+                body: Some(content.clone().into()),
+                content_type: Some(mime.clone()),
+                ..Default::default()
+            }).map_err(move |e| {
+                log::error!("failed to upload to {}: {:?}", bucket_path, e);
+                // try two additional times (for a total of 3)
+                client.put_object(PutObjectRequest {
+                    bucket: "rust-docs-rs".into(),
+                    key: bucket_path.clone(),
+                    body: Some(content.clone().into()),
+                    content_type: Some(mime.clone()),
+                    ..Default::default()
+                }).map_err(|_| {
+                    panic!("failed to upload twice, exiting");
+                })
+            }));
+
+            if futures.len() > 1000 {
```

Comment:
@Mark-Simulacrum something like this?

Reply:
Yep, this is almost perfect -- let's extract the constant into `const MAX_CONCURRENT_UPLOADS: usize = 1000;` at top of the module (after all the imports, though).